### PR TITLE
Various improvements

### PR DESCRIPTION
**Pull request by tkoolen**
On top of #7.

Some additional functionality I needed. See commit messages. No tests yet.

### CODE REVIEW

File: src/MultilinearOpt.jl [hunk 1]
```diff
@@ -230,7 +236,14 @@ function linearize_quadratic!(m::JuMP.Model, t::JuMP.QuadExpr, product_dict::Dic
     aff = copy(t.aff)
     for (coeff, x, y) in JuMP.quad_terms(t)
         z = get!(product_dict, JuMP.UnorderedPair(x, y)) do
-            @assert x != y # TODO: support non-bilinear terms
+            # TODO: better relaxation for the case x == y
+            if JuMP.is_fixed(x) && JuMP.is_fixed(y)
```
Comment (joehuchette) [inline src/MultilinearOpt.jl]: 👍

File: test/constraint_violation.jl [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+        method::Symbol, disc_level::Integer, rng::AbstractRNG, atol::Real, num_tests::Integer)
+    m = Model(with_optimizer(Gurobi.Optimizer, OutputFlag=0))
+    @variable m x lower_bound=xbounds[1] upper_bound=xbounds[2]
+    @variable m y lower_bound=ybounds[1] upper_bound=ybounds[2]
+    @variable m z
+    @constraint m z == f(x, y)
+    relaxbilinear!(m, method=method, disc_level=disc_level)
+    for i = 1 : num_tests
+        x̄ = xbounds[1] + rand(rng) * (xbounds[2] - xbounds[1])
+        ȳ = xbounds[1] + rand(rng) * (xbounds[2] - xbounds[1])
```
Comment (joehuchette) [inline test/constraint_violation.jl]: Should be updated with changes to #7

File: test/constraint_violation.jl [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+    atol = 1.1e-2
+    num_tests = 20
+    xbounds = (-0.5, 0.5)
+    ybounds = (-0.4, 0.6)
+    for (method, disc_level) in  [
+            (:Logarithmic1D, 31),
+            (:Logarithmic2D, 6),
+            (:ZigZag1D, 31),
+            (:ZigZag2D, 6),
+            # (:Unary, 31) # broken
```
Comment (joehuchette) [inline test/constraint_violation.jl]: How so?

**Review by joehuchette — APPROVED** [review]

File: test/constraint_violation.jl (hunk 3 above)
Comment (tkoolen) [reply test/constraint_violation.jl]: To clarify: this test was added in #7 and was already broken before any changes in this PR. I'm getting:

```julia
constraint violation: Error During Test at /home/twan/.julia/dev/MultilinearOpt/test/constraint_violation.jl:33
  Got exception outside of a @test
  UndefVarError: sos2_mc_formulation! not defined
  Stacktrace:
   [1] outerapproximate(::Model, ::Tuple{VariableRef,VariableRef}, ::MultilinearFunction{2}, ::Discretization{2}, ::Symbol) at /home/twan/.julia/dev/MultilinearOpt/src/MultilinearOpt.jl:57
   [2] (::getfield(MultilinearOpt, Symbol("##12#14")){Model,Symbol,Int64,VariableRef,VariableRef})() at /home/twan/.julia/dev/MultilinearOpt/src/MultilinearOpt.jl:242
   [3] get!(::getfield(MultilinearOpt, Symbol("##12#14")){Model,Symbol,Int64,VariableRef,VariableRef}, ::Dict{UnorderedPair{VariableRef},VariableRef}, ::UnorderedPair{VariableRef}) at ./dict.jl:453
   [4] linearize_quadratic!(::Model, ::GenericQuadExpr{Float64,VariableRef}, ::Dict{UnorderedPair{VariableRef},VariableRef}, ::Symbol, ::Int64) at /home/twan/.julia/dev/MultilinearOpt/src/MultilinearOpt.jl:231
   [5] #relaxbilinear!#11(::Symbol, ::Int64, ::Function, ::Model) at /home/twan/.julia/dev/MultilinearOpt/src/MultilinearOpt.jl:210
   [6] #relaxbilinear! at ./none:0 [inlined]
   [7] #test_constraint_violation#3(::Symbol, ::Int64, ::MersenneTwister, ::Float64, ::Int64, ::Function, ::typeof(*), ::Tuple{Float64,Float64}, ::Tuple{Float64,Float64}) at /home/twan/.julia/dev/MultilinearOpt/test/constraint_violation.jl:15
   [8] (::getfield(Main, Symbol("#kw##test_constraint_violation")))(::NamedTuple{(:method, :disc_level, :atol, :rng, :num_tests),Tuple{Symbol,Int64,Float64,MersenneTwister,Int64}}, ::typeof(test_constraint_violation), ::Function, ::Tuple{Float64,Float64}, ::Tuple{Float64,Float64}) at ./none:0
   [9] top-level scope at /home/twan/.julia/dev/MultilinearOpt/test/constraint_violation.jl:49
   [10] top-level scope at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.1/Test/src/Test.jl:1083
   [11] top-level scope at /home/twan/.julia/dev/MultilinearOpt/test/constraint_violation.jl:34
   [12] include at ./boot.jl:326 [inlined]
   [13] include_relative(::Module, ::String) at ./loading.jl:1038
   [14] include(::Module, ::String) at ./sysimg.jl:29
   [15] include(::String) at ./client.jl:403
   [16] top-level scope at none:0
   [17] include at ./boot.jl:326 [inlined]
   [18] include_relative(::Module, ::String) at ./loading.jl:1038
   [19] include(::Module, ::String) at ./sysimg.jl:29
   [20] include(::String) at ./client.jl:403
   [21] top-level scope at none:0
   [22] eval(::Module, ::Any) at ./boot.jl:328
   [23] exec_options(::Base.JLOptions) at ./client.jl:243
   [24] _start() at ./client.jl:436
```

Comment (joehuchette) [reply test/constraint_violation.jl]: Got it, just was curious if the problem was here or in PiecewiseLinearOpt

**Pull request merged**